### ISSUE TITLE
Concat multiple function files

### ISSUE THREAD

**Issue by majoraze** [serverless/serverless#4218]
Hey fellas,

I have a base code with some legacy apis, and I’m trying to unify all in one container repository (with git subrepo) and creating one unique serverless.yml file in this container git.

I’m trying to do the following in my serverless file:
```
functions: 
  ${file(api-1/functions.yml)}
  ${file(api-2/functions.yml)}
```
Inside this functions.yml is simply the function declaration like so:
```
users:
  DependsOn: RedisInstance
  handler: api-1/handler.index
  timeout: 30
  vpc:
    securityGroupIds:
      - Ref: RedisSecurityGroup
    subnetIds:
      - Ref: PrivateSubnet
  events:
    - http: any {proxy+}
  environment:
    stage: ${opt:stage}
```
And I'm getting the error below when I try to do that:

> Serverless Error ---------------------------------------
> 
> Trying to populate non string value into a string for variable ${file(api-2/functions.yml)}. Please make sure the value of the property is a string.

But when I remove one of the file its working as expected, uploading the functions of this file:
```
functions: 
  ${file(api-1/functions.yml)}
```

Thanks!

**Comment by kbariotis** [comment]
@majoraze concatenation of external files in such way is not currently supported. For a quick workaround, I would suggest you to create and include a [JS file](https://serverless.com/framework/docs/providers/aws/guide/variables#reference-variables-in-javascript-files) that will handle the concatenation inside and you will be able to have your configuration in multiple files.

**Comment by dwelch2344** [comment]
We're interested in doing this as well. We naturally started down the JS file route, but ran into questions about how to get access to context-related information so we can do the equivalent of: 
```yml
getAll:
  environment: ${file(env.yml):${self:custom.stage}}
  handler: handler.getAll
  events:
    - http:
        cors: true
        path: users/
        method: get        
```

Is there a way to get access to the existing infrastructure for variable resolution? Or perhaps even just passing that information in to our functions?

**Comment by dwelch2344** [comment]
**TL;DR:** We got it working, example shown below. Ultimately our `serverless-dynamic.js` file loads and combines any root YML props from everything in `./handlers/*.yml`

Would love to roll this into a plugin, if you all think there's demand for it.

---

Okay, a little more tinkering shows this is not necessary as ultimately you're just forming the raw definitions you would normally pass to Serverless. 

So, we found our solution very easily with the following:

**Project Structure**
```
- env.yml
- package.json
- serverless.yml
- serverless-dynamic.js
- src
   - [real source omitted] 
- handlers
   - users.yml
   - projects.yml 
```

**Serverless.yml**
```yml
# NOTE: update this with your service name
service: demo-project

# Use the serverless-webpack plugin to transpile ES6
plugins:
  - serverless-webpack
  - serverless-offline


# Enable auto-packing of external modules
custom:
  webpackIncludeModules: true
  stage: "${opt:stage, self:provider.stage}"

provider:
  name: aws
  runtime: nodejs6.10
  stage: dev
  region: us-east-1

functions:  ${file(serverless-dynamic.js)}
```

**serverless-dynamic.js**
```js 
// requires yamljs:  `npm install --save-dev yamljs`
module.exports = () => {  
  const fs = require('fs')
  const files = fs.readdirSync('./handlers')
  const YAML = require('yamljs')  

  const merged = files
      .map(f => fs.readFileSync(`./handlers/${f}`, 'utf8'))
      .map(raw => YAML.parse(raw))
      .reduce( (result, handler) => Object.assign(result, handler), {})

  return merged
}
```

**users.yml**
```yml
usersGetall:
  environment: ${file(env.yml):${self:custom.stage}}
  handler: handler.usersGetAll
  events:
    - http:
        cors: true
        path: users/
        method: get
```

**projects.yml**
```yml
projectsGetall:
  environment: ${file(env.yml):${self:custom.stage}}
  handler: handler.projectsGetAll
  events:
    - http:
        cors: true
        path: projects/
        method: get
```

**End Result:**
Both functions show up as expected, and the project feels mighty clean :)

**Comment by rkdavidson** [comment]
Thank you @dwelch2344 — your solve steered me right down the path I was looking for.

I was looking for a way to exclude local, offline only lambda functions that make local dynmodb management easier from our non-local stages.
Googling "serverless exclude lambdas from certain stages" or "serverless deploy lambda for certain stage only" etc was too vague/not-helpful.

I adapted your approach as following. Fairly happy with the result aside from inability to figure out how to specify a default/fallback value.

**serverless.yml**
```
...

custom:
  stage: ${opt:stage, self:provider.stage}

functions: ${file(serverless-functions.js):${self:custom.stage}}

...
```

**serverless-functions.js**
```
// Utilities
const fs = require('fs');
const YAML = require('yamljs');

// Attempts to read a local YML file's contents, automatically parsing and return result
function getYmlContents(file) {
  if (fs.existsSync(file)) {
    return YAML.parse(fs.readFileSync(file, 'utf8'));
  } else {
    console.log(`( serverless-functions.js ) Error: File ${file} not found.`);
    return {};
  }
}

// Common function definitions across all stages
const baseFunctions = getYmlContents('./serverless.functions.base.yml');
const extendBaseFunctionsWithYml = (file) => Object.assign({}, baseFunctions, getYmlContents(file));

// Specific environment stages here
module.exports.local = () => extendBaseFunctionsWithYml('./serverless.functions.local.yml');
module.exports.dev = () => baseFunctions;
module.exports.qa = () => baseFunctions;
module.exports.prod = () => baseFunctions;
```

**serverless.functions.base.yml**
```
usersGetall:
  handler: handler.usersGetAll
  events:
    - http:
        cors: true
        path: users/
        method: get
```

**serverless.functions.local.yml**
```
localCreateUser:
  handler: handler.localCreateUser
  events:
    - http:
        cors: true
        path: local/users/
        method: get
```

When I run serverless locally with `-s local`, the functions declaration becomes:
``` 
usersGetall:
  handler: handler.usersGetAll
  events:
    - http:
        cors: true
        path: users/
        method: get

localCreateUser:
  handler: handler.localCreateUser
  events:
    - http:
        cors: true
        path: local/users/
        method: get
```

But when run with a stage like `-s dev`, functions declaration becomes:
``` 
usersGetall:
  handler: handler.usersGetAll
  events:
    - http:
        cors: true
        path: users/
        method: get
```

**Comment by c24w** [comment]
Interestingly, Google only found this issue and not the actual [documentation](https://serverless.com/framework/docs/providers/aws/guide/functions/) describing how to do this now:

>You can specify an array of functions, which is useful if you separate your functions in to different files:
```yml
# serverless.yml
...

functions:
  - ${file(../foo-functions.yml)}
  - ${file(../bar-functions.yml)}
```
```yml
# foo-functions.yml
getFoo:
  handler: handler.foo
deleteFoo:
  handler: handler.foo
```

**Comment by dschep** [comment]
You are correct @c24w!

**Comment by epoleacov** [comment]
> Should this be closed? The [documentation](https://serverless.com/framework/docs/providers/aws/guide/functions/) describes how to do this now:
> 
> > You can specify an array of functions, which is useful if you separate your functions in to different files:
> 
> ```yaml
> # serverless.yml
> ...
> 
> functions:
>   - ${file(../foo-functions.yml)}
>   - ${file(../bar-functions.yml)}
> ```
> 
> ```yaml
> # foo-functions.yml
> getFoo:
>   handler: handler.foo
> deleteFoo:
>   handler: handler.foo
> ```

In this case handler path is wrong and SLS can't find it, so idea of this was to keep handler in same directory with YAML file, bu is not working as expected.

**Comment by c24w** [comment]
> > Should this be closed? The [documentation](https://serverless.com/framework/docs/providers/aws/guide/functions/) describes how to do this now:
> > > You can specify an array of functions, which is useful if you separate your functions in to different files:
> > 
> > 
> > ```yaml
> > # serverless.yml
> > ...
> > 
> > functions:
> >   - ${file(../foo-functions.yml)}
> >   - ${file(../bar-functions.yml)}
> > ```
> > 
> > 
> > ```yaml
> > # foo-functions.yml
> > getFoo:
> >   handler: handler.foo
> > deleteFoo:
> >   handler: handler.foo
> > ```
> 
> In this case handler path is wrong and SLS can't find it, so idea of this was to keep handler in same directory with YAML file, bu is not working as expected.

I can't see anywhere that implies handler paths should be relative to the yml files, rather than the root `serverless.yml`.  For example, see @majoraze's [original post](https://github.com/serverless/serverless/issues/4218#issue-255617750):

> ```yml
> functions: 
>   ${file(api-1/functions.yml)}
> ```
> ```yml
> users:
>   handler: api-1/handler.index
> ```

I assume serverless is essentially just plopping the contents of `foo-functions.yml` and `bar-functions.yml` into the `functions` section of `serverless.yml`.  I think it would be too much to expect it to re-resolve the handler paths at the same time.

**Comment by sourabhv** [comment]
> Should this be closed? The [documentation](https://serverless.com/framework/docs/providers/aws/guide/functions/) describes how to do this now:
> 
> > You can specify an array of functions, which is useful if you separate your functions in to different files:
> 
> ```yaml
> # serverless.yml
> ...
> 
> functions:
>   - ${file(../foo-functions.yml)}
>   - ${file(../bar-functions.yml)}
> ```
> 
> ```yaml
> # foo-functions.yml
> getFoo:
>   handler: handler.foo
> deleteFoo:
>   handler: handler.foo
> ```

If I run `sls print` with this, it prints something like this

```
...
functions:
  - getFoo:
      handler: handler.foo
    deleteFoo:
      handler: handler.foo
```

instead of what I expected: 

```
functions:
  getFoo:
    handler: handler.foo
  deleteFoo:
    handler: handler.foo
```

**Comment by mp3il** [comment]
> > Should this be closed? The [documentation](https://serverless.com/framework/docs/providers/aws/guide/functions/) describes how to do this now:
> > > You can specify an array of functions, which is useful if you separate your functions in to different files:
> > 
> > 
> > ```yaml
> > # serverless.yml
> > ...
> > 
> > functions:
> >   - ${file(../foo-functions.yml)}
> >   - ${file(../bar-functions.yml)}
> > ```
> > 
> > 
> > ```yaml
> > # foo-functions.yml
> > getFoo:
> >   handler: handler.foo
> > deleteFoo:
> >   handler: handler.foo
> > ```
> 
> If I run `sls print` with this, it prints something like this
> 
> ```
> ...
> functions:
>   - getFoo:
>       handler: handler.foo
>     deleteFoo:
>       handler: handler.foo
> ```
> 
> instead of what I expected:
> 
> ```
> functions:
>   getFoo:
>     handler: handler.foo
>   deleteFoo:
>     handler: handler.foo
> ```

Did you manage to solve it? What is the best practice for splitting functions into multiple files? and later including them in serverless.yml ?

**Comment by dzyubadmytro** [comment]
@mp3il It worked perfect for me. But didnt work when I tried to do something like this 
```
functions:
    - test:
        handler: handler.foo
    - ${file(../foo-functions.yml)}
    - ${file(../bar-functions.yml)}
```
Also make sure that _emitDeclarationOnly_, property in tsconfig.json is **not** set to _true_.

**Comment by Schachte** [comment]
> 

I think I experienced the same behavior when doing this with `Resources` block as well.
